### PR TITLE
Add real-time analytics support

### PR DESCRIPTION
**Pull request by larshelge**
The main goal of this PR is to make data more quickly available in analytics tables. The overall approach is to introduce a new analytics table partition which holds only the data captured since last full analytics table update, and always include this partition in analytics queries.

The PR does the following:

* Introduces a new partition table called "latest". This holds the latest data, i.e. data added or updated _since last full analytics table generation_.  The latest partition uses a "year" value of 0.
* The latest partition is dropped and rebuilt on each run (like the regular analytics partitions).
* For each analytics table run, the system checks whether there is new, updated or deleted data since last run.
* When considering whether there is new data, the timestamp of the _start_ of this process is used to guarantee no overlap between runs.
* Data which is updated or deleted through runs is deleted from any of the analytics partitions (including old data).
* A new syntetic unique identifier column is added to the analytics tables for the purpose of deleting updated/deleted data.
* The system checks whether new data exists since last update of the latest partition, and if not, aborts the update.

The scheduling part is yet to be done. Triggering the "latest" partition can be done through the API like this:

    POST /api/31/resourceTables/analytics?lastYears=0&skipResourceTables=true

### CODE REVIEW

**Comment by jason-p-pickering** [comment]
Can one of the admins verify this patch?

File: dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/AnalyticsTablePartition.java [hunk 1]
```diff
@@ -111,6 +114,11 @@ public String getTempTableName()
         return name;
     }
 
+    public boolean isLatestPartition()
+    {
+        return year == LATEST_PARTITION;
```
Comment (luciano-fiandesio) [inline dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/AnalyticsTablePartition.java]: I would use `return Objects.equals(year, LATEST_PARTITION);` or `equals(..)`

File: dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/AnalyticsTableUpdateParams.java [hunk 2]
```diff
@@ -108,14 +114,27 @@ public Date getStartTime()
...
-        return lastYears != null;
+        return lastYears != null || isLatestUpdate();
+    }
+
+    /**
+     * Indicates whether this is an update of the "latest" partition.
+     */
+    public boolean isLatestUpdate()
+    {
+        return lastYears == AnalyticsTablePartition.LATEST_PARTITION;
```
Comment (luciano-fiandesio) [inline dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/AnalyticsTableUpdateParams.java]: I would use `return Objects.equals(year, LATEST_PARTITION);` or `equals(..)`

File: dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/table/AbstractEventJdbcTableManager.java [hunk 3]
```diff
@@ -149,6 +150,12 @@ public String validState()
         return null;
     }
 
+    @Override
+    protected boolean hasUpdatedLatestData( Date startDate, Date endDate )
+    {
+        throw new IllegalStateException( "Should not be invoked for event table managers" );
```
Comment (luciano-fiandesio) [inline dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/table/AbstractEventJdbcTableManager.java]: `AnalyticsTableType` has EVENT and ENROLLMENT with the property `latestPartition` set to `true`. Maybe I misunderstood?

File: dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/AnalyticsTablePartition.java (hunk 1 above)
Comment (larshelge) [reply dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/AnalyticsTablePartition.java]: Good catch, fixed.

File: dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/AnalyticsTableUpdateParams.java (hunk 2 above)
Comment (larshelge) [reply dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/AnalyticsTableUpdateParams.java]: Agree, fixed.

File: dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/table/AbstractEventJdbcTableManager.java (hunk 3 above)
Comment (larshelge) [reply dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/table/AbstractEventJdbcTableManager.java]: Good point, in fact ENROLLMENT should have latestPartition = false. Updated that now.

Comment (larshelge) [reply dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/table/AbstractEventJdbcTableManager.java]: Agree on returning false vs exception too. Changed it to false.

**Pull request merged**